### PR TITLE
CA-220047: Restored error message stating that VM prevents host evacu…

### PR DESCRIPTION
**Pull request by kc284**
…ation due to missing PV drivers.

The generic message VM_LACKS_FEATURE_SUSPEND does not offer much detail to the
user.

Signed-off-by: Konstantina Chremmou <konstantina.chremmou@citrix.com>

### CODE REVIEW

**Comment by kc284** [comment]
@thomassa as you worked on CP-9366 could you please review this?

**Comment by kc284** [comment]
Marking this as needs updating till I figure out the test fialures.

File: ocaml/xapi/xapi_vm_lifecycle.ml [hunk 1]
```diff
@@ -150,15 +150,16 @@ let has_feature ~vmgmr ~feature =
...
  *  react helpfully. *)
 let check_op_for_feature ~__context ~vmr ~vmgmr ~power_state ~op ~ref ~strict =
-  if power_state <> `Running ||
-     (* PV guests offer support implicitly *)
-     not (Helpers.has_booted_hvm_of_record ~__context vmr) ||
-     Xapi_pv_driver_version.(has_pv_drivers (of_guest_metrics vmgmr)) (* Full PV drivers imply all features *)
-  then None
+  let some_err e = Some (e, [ Ref.string_of ref ]) in
+
+  if power_state <> `Running then None
```
Comment (mseri) [inline ocaml/xapi/xapi_vm_lifecycle.ml]: I think we need to leave `not (Helpers.has_booted_hvm_of_record ~__context vmr) || Xapi_pv_driver_version.(has_pv_drivers (of_guest_metrics vmgmr))` in the condition

**Comment by mseri** [comment]
marking as blocked until we unfreeze trunk

**Comment by thomassa** [comment]
I'm not keen on this. The `check_op_for_feature` function is to see whether the guest has the feature(s) that are needed for us to perform the specified operation on it.

The new bit added into that function in this pull-request is
```
+  else if (Helpers.has_booted_hvm_of_record ~__context vmr) &&
+    not Xapi_pv_driver_version.(has_pv_drivers (of_guest_metrics vmgmr))
+  then some_err Api_errors.vm_missing_pv_drivers
```
i.e. we raise an error if the guest has not reported PV driver version.
But in principle a guest can have all the features but not report a PV driver version.
I have a feeling this might be normal for HVM Linux guests, though I'm not sure.

**Comment by pauldu** [comment]
Quite right. Nothing should be checking for a monolithic 'PV driver version'... it doesn't exist now that we post on Windows Update.

**Comment by kc284** [comment]
@thomassa would it be then more correct to move the check further down and show this error as a subcase of the lacks feature errors?
Presumably the concern on the ui side is that the user may want to know in more detail why the guest lacks the feature.

**Comment by thomassa** [comment]
In theory a guest could have a PV driver version but not have some particular feature, though that's not likely ever to happen.

But really, the error is that the guest lacks the feature which is needed for the operation we want to perform.
In some cases this might be remediable by installing a package of PV drivers, but I feel that sort of interpretation of error-messages should be done client-side.

**Comment by jonludlam** [comment]
I think this particular PR is a step backwards - we _may_ need to do something to make the error better, but if we do it'll look substantially different than this PR. Hence I'm closing it.

**Pull request closed without merging**